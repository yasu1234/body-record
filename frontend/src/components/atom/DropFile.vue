<template>
  <div class="main">
    <div
      class="dropzone-container"
      @dragover="dragover"
      @dragleave="dragleave"
      @drop="drop"
    >
      <input
        type="file"
        name="file"
        :id="`fileInput-${index}`"
        class="hidden-input"
        @change="onChange"
        @click="
          (e) => {
            e.target.value = '';
          }
        "
        accept=".pdf,.jpg,.jpeg,.png"
      />
      <label :for="`fileInput-${index}`" class="file-label">
        <div v-if="isDragging">Release to drop files here.</div>
        <div v-else>
          ファイルをドロップ、もしくは<u>こちらをクリックして</u>アップロードしてください
        </div>
      </label>
      <div class="preview-container mt-4" v-if="file !== null">
        <div :key="index" class="preview-card">
          <div>
            <img class="preview-img" :src="generateThumbnail(file)" />
            <p :title="file.name">{{ file.name }}</p>
          </div>
          <div>
            <button
              class="delete-button"
              @click="remove"
              title="Remove file"
            >
              <b>&times;</b>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup>
import { ref, onMounted } from "vue";

const isDragging = ref(false);
const index = ref(0);
const file = ref(null);

const props = defineProps({
  index: Number,
});

const emit = defineEmits(["change"]);

onMounted(() => {
  index.value = props.index;
});

const onChange = (e) => {
  file.value = e.target.files[0];
  emit("change", file.value, index.value);
};

const generateThumbnail = (file) => {
  let fileSrc = URL.createObjectURL(file);
  setTimeout(() => {
    URL.revokeObjectURL(fileSrc);
  }, 1000);
  return fileSrc;
};

const remove = () => {
  file.value = null;
  emit("change", file.value, index.value);
};

const dragover = (e) => {
  e.preventDefault();
  isDragging.value = true;
};

const dragleave = () => {
  isDragging.value = false;
};

const drop = (e) => {
  e.preventDefault();
  file.value = e.dataTransfer.files[0];
  isDragging.value = false;
  emit("change", file.value, index.value);
};
</script>

<style scoped>
.main {
  width: 400px;
  text-align: center;
}
.dropzone-container {
  padding: 3rem;
  border: 1px solid #343536;
}
.hidden-input {
  opacity: 0;
  overflow: hidden;
  position: absolute;
  width: 1px;
  height: 1px;
}
.file-label {
  font-size: 20px;
  display: block;
  cursor: pointer;
}
.preview-container {
  display: flex;
  margin-top: 2rem;
}
.preview-card {
  display: flex;
  border: 1px solid #a2a2a2;
  padding: 5px;
  margin-left: 5px;
}
.preview-img {
  width: 100px;
  height: 100px;
  border-radius: 5px;
  border: 1px solid #a2a2a2;
}
.delete-button {
  width: 20px;
  height: 20px;
  border-radius: 50%;
}

@media screen and (max-width: 768px) {
  .main {
    width: auto;
    text-align: center;
    margin-right: 20px;
  }
}
</style>
