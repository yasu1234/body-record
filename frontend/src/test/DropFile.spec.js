import { shallowMount } from "@vue/test-utils";
import DropFile from "@/components/atom/DropFile.vue";

jest.mock("axios");
jest.mock("vue-router", () => {
  const router = {
    push: jest.fn(),
    pathname: "/",
    route: "",
    asPath: "",
    query: {},
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
  };
});

jest.mock("primevue/usetoast", () => ({
  useToast: jest.fn(() => ({
    add: jest.fn(),
    removeAll: jest.fn(),
  })),
}));

describe("DropFile", () => {
  let wrapper;

  beforeEach(() => {
    global.URL.createObjectURL = jest.fn(() => 'mock-object-url');
    global.URL.revokeObjectURL = jest.fn();
    wrapper = shallowMount(DropFile, {
      props: {
        index: 0,
      },
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("ファイルが選択された時に、changeイベントが正しく発行される", async () => {
    const file = new File(["hello"], "hello.jpg", { type: "img/jpg" });
    const input = wrapper.find('input[type="file"]');
    Object.defineProperty(input.element, "files", {
      value: [file],
      writable: false,
    });
    await input.trigger("change");
    expect(wrapper.emitted("change")[0]).toEqual([file, 0]);
  });

  it("ファイルがドロップされた時に、ファイルデータが正しく設定される", async () => {
    const file = new File(["hello"], "hello.txt", { type: "text/plain" });
    const dropzone = wrapper.find(".dropzone-container");
    await dropzone.trigger("drop", { dataTransfer: { files: [file] } });
    expect(wrapper.vm.file).toEqual(file);
  });

  it("削除ボタンがクリックされた時に、ファイルデータが正しくリセットされ、change イベントが正しく発行される", async () => {
    const file = new File(["hello"], "hello.txt", { type: "text/plain" });
    wrapper.vm.file = file;
    await wrapper.vm.$nextTick();
    const removeButton = wrapper.find("button");
    await removeButton.trigger("click");
    expect(wrapper.vm.file).toBeNull();
    expect(wrapper.emitted("change")[0]).toEqual([null, 0]);
  });
});
