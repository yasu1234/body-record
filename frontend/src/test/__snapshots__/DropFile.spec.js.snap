// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropFile SnapShotテスト初期表示 1`] = `
<div
  class="main"
>
  <div
    class="dropzone-container"
  >
    <input
      accept=".pdf,.jpg,.jpeg,.png"
      class="hidden-input"
      id="fileInput-0"
      name="file"
      type="file"
    />
    <label
      class="file-label"
      for="fileInput-0"
    >
      <div>
         ファイルをドロップ、もしくは
        <u>
          こちらをクリックして
        </u>
        アップロードしてください 
      </div>
    </label>
    <!--v-if-->
  </div>
</div>
`;
